common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java

```diff
@@ -648,14 +809,24 @@ public void onData(String streamId, ByteBuffer buf) throws IOException {
       // to disk as well. This way, we avoid having to buffer the entirety of every blocks in
       // memory, while still providing the necessary guarantee.
       synchronized (partitionInfo) {
-        Map<Integer, AppShufflePartitionInfo> shufflePartitions =
+        Map<Integer, Map<Integer, AppShufflePartitionInfo>> shuffleMergePartitions =
           appShuffleInfo.partitions.get(partitionInfo.shuffleId);
-        // If the partitionInfo corresponding to (appId, shuffleId, reduceId) is no longer present
-        // then it means that the shuffle merge has already been finalized. We should thus ignore
-        // the data and just drain the remaining bytes of this message. This check should be
-        // placed inside the synchronized block to make sure that checking the key is still
-        // present and processing the data is atomic.
-        if (shufflePartitions == null || !shufflePartitions.containsKey(partitionInfo.reduceId)) {
+        // Older shuffleMergePartitions gets cleaned up when newer shuffleMergeId gets created
```

Comment:
I doubt this. IIUC, when there's a newer `shuffleMergeId`, we'd set `shuffleMergePartitions(olderShuffleMergeId)=STALE_SHUFFLE_PARTITIONS` rather than clean up the  `shuffleMergePartitions` from `appShuffleInfo.partitions`.

Reply:
This is how it goes for indeterminate stages.
1. `shuffleMergeId` starts with 1.
2. When `shuffleMergeId` moves to 2 `shuffleMergePartitions(1) = STALE_SHUFFLE_PARTITIONS`. 
3. Lets say now `shuffleMergeId` moves to 3 then `shuffleMergePartitions(2) = STALE_SHUFFLE_PARTITIONS` and `shuffleMergePartitions(1)` will be removed.

So we keep the last 2 recent entries of `shuffleMergeId` in the `shuffleMergePartitionsMap` Do you want me to clarify the comment further?

Reply:
> 3. Lets say now shuffleMergeId moves to 3 then shuffleMergePartitions(2) = STALE_SHUFFLE_PARTITIONS and shuffleMergePartitions(1) will be removed.

We don't clean the `shuffleMergePartitions` but its entry(e.g., shuffleMergePartitions(1)), right?

What I doubt is that `shuffleMergePartitions` shouldn't be null but `shuffleMergePartitions(shuffleMergeId)` can be null.

Reply:
Yes thats correct, will clarify the comment.

Reply:
We should not only clarify the comment but also remove this check (`shuffleMergePartitions == null`). IIUC, it would never happen. Or maybe, replace it with an assertion (assert shuffleMergePartitions != null.)

Reply:
Refactored this piece of code again. Now I'm cleaning up the metadata for determinate stages as before so shuffleMergePartitions can be null for those shuffleIds. Please take a look. Thanks!

Reply:
Refactored this further to get rid of the complex Map by introducing a new object `AppShuffleMergePartitionsInfo` containing `shuffleMergeId` and `Map with the shuffle partitions`